google-cloud-firestore/src/main/java/com/google/cloud/firestore/TransactionOptions.java

```diff
@@ -16,65 +16,98 @@
 
 package com.google.cloud.firestore;
 
+import com.google.api.core.InternalApi;
 import com.google.common.base.Preconditions;
+import com.google.firestore.v1.TransactionOptions.ReadOnly;
+import com.google.firestore.v1.TransactionOptions.ReadWrite;
+import com.google.protobuf.Timestamp;
+import com.google.protobuf.TimestampOrBuilder;
 import java.util.concurrent.Executor;
 import javax.annotation.Nonnull;
 import javax.annotation.Nullable;
 
 /** Options specifying the behavior of Firestore Transactions. */
 public final class TransactionOptions {
 
+  private static final TransactionOptions DEFAULT_READ_WRITE_TRANSACTION_OPTIONS =
+      readWriteOptionsBuilder().build();
+
   private static final int DEFAULT_NUM_ATTEMPTS = 5;
 
-  private final int numberOfAttempts;
   private final Executor executor;
+  private final EitherReadOnlyOrReadWrite options;
 
-  TransactionOptions(int maxAttempts, Executor executor) {
-    this.numberOfAttempts = maxAttempts;
+  TransactionOptions(Executor executor, EitherReadOnlyOrReadWrite options) {
     this.executor = executor;
+    this.options = options;
   }
 
+  /**
```

Comment:
Read-Only transactions can fail with ABORTED too (if a read-write transaction holds a lock) and hence I think it still makes sense to retry.

Please correct me if I am wrong, since I would think that you know more about this than I do.

Reply:
Based on the lack of a `retry_transaction` like [`ReadWrite`](https://github.com/googleapis/java-firestore/blob/0c841f9b8c03aaeceb2762492daeeda9f8efdd66/proto-google-cloud-firestore-v1/src/main/proto/google/firestore/v1/common.proto#L59-L62), I interpreted it as not safe to retry and instead would require the application layer to try again. Also if a `readTime` is specified it is not allowed to be more than 60 seconds ago from the time it hits the server, so if multiple retries were attempted (especially if the timeframe was smaller than 60 seconds) it's possible the specified timestamp would fallout of the valid time frame.

Reply:
1. I am trying to get this SDK to take over some of the "standard retry" work that otherwise lives in every customer's application layer and ideally would offer a product that just works and irons out what I (personally) consider quirks of the backend.

2. This kind of saddens me but I think that is a valid point and means we should not offer retries.